common/PriorityCache: Updated Implementation of Cache Age Binning

markhpc opened the pull request:
Updated version of the age binning code.  Functions and appears to improve cache behavior but performance tests are showing little gains right now (significantly more testing is needed).

Signed-off-by: Mark Nelson <mnelson@redhat.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

markhpc commented:
So ultimately this PR does what it was intended to do as far as I can tell.  We do a much better job of favoring hot caches than cold caches as can be seen in various workload scenarios here:

https://docs.google.com/spreadsheets/d/1lSp2cLzYmRfPILDCyLMXciIfdf0OvSFngwXukQFXIqQ/edit?usp=sharing

The bad news is that overall it's basically a draw from a performance and CPU usage perspective when testing on our very fast Mako nodes.  It's possible that further tuning may show additional differences, but one of the primary lessons from developing this code is that caching OMAP in the rocksdb block cache is much less useful than I expected.  For instance in RGW, one of the only common use cases where having cached omap appears to be helpful is during bucket index listing.  Since that involves doing iterative scans, the far more important thing to have is readahead vs a typical LRU style cache.  That may partially explain why it was far better in previous versions of Ceph to give almost all of the cache to bluestore's onode cache rather than dedicating a significant amount to the rocksdb block cache.

There are a couple of cases where cache age binning is slightly faster (primarily when we have repeated data cache reads) but also some cases where it's a little slower (primarily in cases where we cached data on large writes that was seldomly read but forced out semi-cold onodes that later had to be re-read).

At this point the primary benefit of this PR appears to be that it more aggressively trims old onodes leading to less memory fragmentation (almost like background compaction) and provides much better insight into the relative ages of different caches so that we now can better understand and communicate how we are prioritizing memory.  It also gives us much finer grained control over how we prioritize different caches.

One other possibility is that this PR may show a bigger improvement with slower devices, especially with mixed workloads.  On this test system, data cache misses during reads had no noticeable penalty due to the high performance of the underlying device.  At the moment I don't have easy access to a system with HDDs to test on so we have not yet tried that scenario.

markhpc commented:
Ok, reproduced the segfault fairly easily on my test node and traced it back to some shady unprotected accesses to the age bin data structure.  Fixed those and got through a couple of rounds of ceph_test_objectstore without segfaults.  I did hit a couple of failures but I don't think they are likely related.

> [       OK ] ObjectStore/StoreTestOmapUpgrade.LargeLegacyToPG/2 (6056 ms)
> [ RUN      ] ObjectStore/StoreTestOmapUpgrade.LargeLegacyToPG/3
> [       OK ] ObjectStore/StoreTestOmapUpgrade.LargeLegacyToPG/3 (0 ms)
> [----------] 12 tests from ObjectStore/StoreTestOmapUpgrade (15639 ms total)
> 
> [----------] Global test environment tear-down
> [==========] 484 tests from 3 test suites ran. (3788154 ms total)
> [  PASSED  ] 478 tests.
> [  FAILED  ] 6 tests, listed below:
> [  FAILED  ] ObjectStore/StoreTest.CompressionTest/2, where GetParam() = "bluestore"
> [  FAILED  ] ObjectStore/StoreTest.ZeroLengthZero/3, where GetParam() = "kstore"
> [  FAILED  ] ObjectStore/StoreTest.OMapTest/3, where GetParam() = "kstore"
> [  FAILED  ] ObjectStore/StoreTest.OMapIterator/3, where GetParam() = "kstore"
> [  FAILED  ] ObjectStore/StoreTestSpecificAUSize.BluestoreStatFSTest/2, where GetParam() = "bluestore"
> [  FAILED  ] ObjectStore/StoreTestSpecificAUSize.garbageCollection/2, where GetParam() = "bluestore"

ljflores commented:
http://pulpito.front.sepia.ceph.com/yuriw-2022-01-04_21:52:15-rados-wip-yuri7-testing-2022-01-04-1159-distro-default-smithi/

Failures, unrelated:
    https://tracker.ceph.com/issues/53723
    https://tracker.ceph.com/issues/38357
    https://tracker.ceph.com/issues/53294
    https://tracker.ceph.com/issues/53424
    https://tracker.ceph.com/issues/53680
    https://tracker.ceph.com/issues/36524
    https://tracker.ceph.com/issues/53781
    
Details:
    Bug_#53723: Cephadm agent fails to report and causes a health timeout - Ceph - Orchestrator
    Bug_#38357: ClsLock.TestExclusiveEphemeralStealEphemeral failed - Ceph - RADOS
    Bug_#53294: rados/test.sh hangs while running LibRadosTwoPoolsPP.TierFlushDuringFlush - Ceph - RADOS
    Bug_#53424: CEPHADM_DAEMON_PLACE_FAIL in orch:cephadm/mgr-nfs-upgrade/ - Ceph - Orchestrator
    Bug_#53680: ERROR:tasks.rook:'waiting for service removal' reached maximum tries (90) after waiting for 900 seconds - Ceph - Orchestrator
    Bug_#36524: hang on 'sudo tar cz -f - -C /var/log/ceph -- .' - Tools - Teuthology
    Bug_#53781: cephadm/test_dashboard_e2e.sh: Unable to find element `cd-modal` when testing on orchestrator/03-inventory.e2e-spec.ts - Ceph - Mgr - Dashboard

ljflores commented:
jenkins retest this please

ljflores commented:
@markhpc @aclamk Tests are passing, but this still needs a final review from Adam.

markhpc commented:
@neha @aclamk if one of you wants to approve we can go for it.  I was initially nervous about the locking issues after the first round of testing but it was more straightforward to diagnose and fix than I was expecting.  Basically an unprotected access to the age-bin along with a locked read/modify/locked write that was completely unnecessary and easily replaced with straight locked write.

We aren't likely to see significant performance wins with this which is disappointing, but it does give us (and power users) much better insight and control into memory allocation behavior.

ljflores commented:
jenkins retest this please

aclamk commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -1229,11 +1237,15 @@ struct LruBufferCacheShard : public BlueStore::BufferCacheShard {
   void _rm(BlueStore::Buffer *b) override {
     ceph_assert(buffer_bytes >= b->length);
     buffer_bytes -= b->length;
+    assert(*(b->cache_age_bin) >= b->length);
```
ceph_assert ?

markhpc replied:
good catch!

aclamk commented on src/common/options/global.yaml.in:
```diff
@@ -4673,6 +4673,70 @@ options:
+    A 10 element, space separated list of cache age intervals grouped by 
+    priority such that PRI1=[0,n), PRI2=[n,n+1), PRI3=[n+1,n+2) ... 
+    PRI10=[n+8,n+9).  Values are specified in seconds. A 0 in the 2nd term
+    will prevent any items from being associated with that bin. For example,
```
Values are not specified in seconds.
Values are specified in bluestore_cache_autotune_interval units.

aclamk replied:
I think it should be in seconds.
We could of course provided dedicated configuration that will clock bins, but I think it would be overkill.

It should not dependent on cache_autotune_interval, since time base for ageing of bins is important, and cache_autotune_interval is just a update frequency.

I think balance() should be invoked as is, but `rotate_bins()` should be invoked 1/s.
I guess it would mean some math on cache_autotune_interval, or is it acceptable that `rotate_bins()` are not invoked precisely?

markhpc replied:
I think the intervals can be independent.  When we balance we just look at whatever the most recently updated state of the caches were.  It doesn't really matter if it was 1 second ago or 4 seconds ago.  The bigger advantage by separating them is that we can allow for much higher resolution on the age bins.  Instead of an autotune_interval minimum bin size, it can now be as small as you want to make it (with 1 second being a nice easy default imho).

aclamk commented on src/common/PriorityCache.h:
```diff
@@ -100,6 +100,19 @@ namespace PriorityCache {
+    virtual void import_intervals(const std::vector<uint64_t> &intervals) = 0;
+
+    // Set intervals (PRI0 and LAST should be ignored)
+    virtual void set_intervals(PriorityCache::Priority pri, uint64_t end_interval) = 0;
```
set_intervals seems to be never used.

aclamk commented on src/kv/rocksdb_cache/BinnedLRUCache.h:
```diff
@@ -296,6 +326,20 @@ class alignas(CACHE_LINE_SIZE) BinnedLRUCacheShard : public CacheShard {
   mutable std::mutex mutex_;
+
+  // Circular buffer of byte counters for age binning
+  boost::circular_buffer<std::shared_ptr<uint64_t>> age_bins;
```
Is this circularity ever used?
We push new element in `rotate_bins` and trim in `import_intervals`.
Maybe it is just a list?

aclamk commented on src/kv/rocksdb_cache/BinnedLRUCache.h:
```diff
@@ -230,6 +232,34 @@ class alignas(CACHE_LINE_SIZE) BinnedLRUCacheShard : public CacheShard {
+
+  // Collect internal stats of shard
+  void add_stats(size_t& capacity,
+		 size_t& high_pri_pool_usage,
```
I do not see implementation for this.

aclamk commented on src/kv/rocksdb_cache/BinnedLRUCache.h:
```diff
@@ -296,6 +326,20 @@ class alignas(CACHE_LINE_SIZE) BinnedLRUCacheShard : public CacheShard {
+  boost::circular_buffer<std::shared_ptr<uint64_t>> age_bins;
+
+  std::atomic<uint64_t> inserts_low = 0;
+  std::atomic<uint64_t> inserts_high = 0;
```
Unused variables. No implementation/use of `add_stats` ?

aclamk commented on src/kv/rocksdb_cache/BinnedLRUCache.cc:
```diff
@@ -460,6 +483,21 @@ size_t BinnedLRUCacheShard::GetPinnedUsage() const {
   return usage_ - lru_usage_;
 }
 
+void BinnedLRUCacheShard::rotate_bins() {
```
Why `rotate`? If it were `age` we would have here the reason for "Age Binning" :D

markhpc replied:
hrm, perhaps.  Age ties the idea more closely to the concept of age binning, but I like in some ways keeping the concept of bin rotation somewhat distinct from the concept of aging.  IE perhaps at some point in the future rotating the bins might not be triggered strictly by an aging event.  Do you feel strongly about changing the wording here?

aclamk replied:
Maybe `shift_bins`?
My opposition to `rotate` is because there is nothing circular in bins.
But I will not insist on it.

markhpc replied:
I can get behind shift_bins.  I'll make the change.

aclamk commented on src/kv/rocksdb_cache/BinnedLRUCache.cc:
```diff
@@ -577,22 +615,33 @@ int64_t BinnedLRUCache::request_cache_bytes(PriorityCache::Priority pri, uint64_
+    {
+      ceph_assert(pri > 0 && pri < PriorityCache::Priority::LAST);
+      auto prev_pri = static_cast<PriorityCache::Priority>(pri - 1);
+      uint64_t start = get_intervals(prev_pri);
```
Cannot if just be `get_intervals(pri - 1)` ?

aclamk reviewed (changes requested):
Nice, simple, functional code.
Cleanup still required.

markhpc commented on src/common/options/global.yaml.in:
```diff
@@ -4673,6 +4673,70 @@ options:
+    priority such that PRI1=[0,n), PRI2=[n,n+1), PRI3=[n+1,n+2) ... 
+    PRI10=[n+8,n+9).  Values are specified in seconds. A 0 in the 2nd term
+    will prevent any items from being associated with that bin. For example,
+    "1 5 10 0 0 0 0 0 0 0" would define two bins. The first bin counts items 
```
0-1 is an implicit bin that always exists, so in reality you are only creating the last two.

markhpc commented:
@neha-ojha Ok, updated, rebased on master, compiles locally.  I'll run some performance tests with https://github.com/ceph/ceph/pull/44538 applied and rerun make check just to dot the is.  Feel free to run this through teuthology as well.

markhpc commented:
local ceph_test_objectstore passed with only failures in kstore.

yuriw commented:
jenkins test api

markhpc commented:
Performance numbers below are from the age binning test suite (using a base ceph version of de40e940 +  PR #44538).  Tests are performed in descending order as shown.  Intial RBD read performance might have dropped a little bit, though subsequent iterations of tests show almost exactly the same numbers as in master.  It was initially hoped that the age binning code would have dramatically improved the behavior during transition switches from rbd to rgw and back, but it turns out that favoring onode cache is almost always the right choice even for RGW (and we tend to do this implicitly as rocksdb compaction appears to often invalidate data in the block cache).  It's possible we may want to revisit the bin<->priority mappings but for now the primary benefit of this PR is gaining more control and insight into how the caches are receiving and utilizing memory.

Prefill
test |  notes
-- | --
rbd | 4 128GB volumes using 4MB writes
rgw | 4 million 4KB objects

Trail 0
test | master | wip-age-binning
-- | -- | --
4K RBD randread | 104628 iops | 100544 iops
4K RBD randwrite | 62239 iops | 62679 iops
4K RGW gets | 3718 gets/s | 3731 gets/s
RGW Bucket List | 155.93 s | 158.53 s

Trail 1
test | master | wip-age-binning
-- | -- | --
4K RBD randread | 98197 iops | 97362 iops
4K RBD randwrite | 61445 iops | 62371 iops
4K RGW gets | 3742 gets/s | 3744 gets/s
RGW Bucket List | 158.63 s | 158.18 s

Trail 2
test | master | wip-age-binning
-- | -- | --
4K RBD randread | 97601 iops | 96828 iops
4K RBD randwrite | 59776 iops | 59992 iops
4K RGW gets | 3742 gets/s | 3719 gets/s
RGW Bucket List | 157.80 s | 158.09 s
